rpcs3/Emu/RSX/Common/texture_cache.h

```diff
@@ -10,328 +11,106 @@ extern u64 get_system_time();
 
 namespace rsx
 {
-	enum texture_create_flags
-	{
-		default_component_order = 0,
-		native_component_order = 1,
-		swapped_native_component_order = 2,
-	};
-
-	enum memory_read_flags
-	{
-		flush_always = 0,
-		flush_once = 1
-	};
-
-	struct typeless_xfer
-	{
-		bool src_is_typeless = false;
-		bool dst_is_typeless = false;
-		bool src_is_depth = false;
-		bool dst_is_depth = false;
-		u32 src_gcm_format = 0;
-		u32 dst_gcm_format = 0;
-		f32 src_scaling_hint = 1.f;
-		f32 dst_scaling_hint = 1.f;
-		texture_upload_context src_context = texture_upload_context::blit_engine_src;
-		texture_upload_context dst_context = texture_upload_context::blit_engine_dst;
-
-		void analyse()
-		{
-			if (src_is_typeless && dst_is_typeless)
-			{
-				if (src_scaling_hint == dst_scaling_hint &&
-					src_scaling_hint != 1.f)
-				{
-					if (src_is_depth == dst_is_depth)
-					{
-						src_is_typeless = dst_is_typeless = false;
-						src_scaling_hint = dst_scaling_hint = 1.f;
-					}
-				}
-			}
-		}
-	};
-
-	struct cached_texture_section : public rsx::buffered_section
+	template <typename commandbuffer_type, typename section_storage_type, typename image_resource_type, typename image_view_type, typename image_storage_type, typename texture_format>
+	class texture_cache : public rsx::texture_cache_base<section_storage_type>
 	{
-		u16 width;
-		u16 height;
-		u16 depth;
-		u16 mipmaps;
+		static_assert(std::is_base_of<rsx::cached_texture_section<section_storage_type>, section_storage_type>::value, "section_storage_type must derive from rsx::cached_texture_section");
 
-		u16 real_pitch;
-		u16 rsx_pitch;
-
-		u32 gcm_format = 0;
-		bool pack_unpack_swap_bytes = false;
-
-		u64 sync_timestamp = 0;
-		bool synchronized = false;
-		bool flushed = false;
-
-		u32 num_writes = 0;
-		std::deque<u32> read_history;
-
-		u64 cache_tag = 0;
-
-		memory_read_flags readback_behaviour = memory_read_flags::flush_once;
-		rsx::texture_create_flags view_flags = rsx::texture_create_flags::default_component_order;
-		rsx::texture_upload_context context = rsx::texture_upload_context::shader_read;
-		rsx::texture_dimension_extended image_type = rsx::texture_dimension_extended::texture_dimension_2d;
+	public:
+		using superclass           = typename rsx::texture_cache_base<section_storage_type>;
```

Comment:
Regarding superclass, it's a confusing term from some microsoft realm. Simply `base` would be much better.

Reply:
I actually used it because of Python/Ruby influences. I don't mind renaming it.